File: c++/6198.cpp
// my solution
// #include <iostream>
// #include <vector>

// using namespace std;

// vector<pair<long long, int>> stack;
// long long ans;

// int main() {
//   freopen("example.txt", "r", stdin);

//   int N;
//   cin >> N;

//   for (int i = 0; i < N; i++) {
//     long long a;
//     cin >> a;

//     while (stack.size() > 0 && stack.back().first <= a) {
//       ans += i - stack.back().second - 1;
//       stack.pop_back();
//     }
//     stack.push_back({a, i});
//   }

//   while (stack.size() != 0) {
//     ans += N - stack.back().second - 1;
//     stack.pop_back();
//   }

//   cout << ans;
//   return 0;
// }

// another solution
#include <iostream>
#include <vector>

using namespace std;

vector<long long> stack;
long long ans;

int main() {
  freopen("example.txt", "r", stdin);

  int N;
  cin >> N;

  for (int i = 0; i < N; i++) {
    long long a;
    cin >> a;

    while (!stack.empty() && stack.back() <= a) {
      stack.pop_back();
    }
    ans += stack.size();
    stack.push_back(a);
  }

  cout << ans;
  return 0;
}

File: js/6198.js
const fs = require('fs');

const inputs = fs
  .readFileSync('example.txt')
  .toString()
  .trim()
  .split('\n')
  .map(Number);

let ans = 0;
const stack = [];

const N = inputs[0];

for (let i = 1; i <= N; i++) {
  while (stack.length > 0 && stack[stack.length - 1][0] <= inputs[i]) {
    const [_, j] = stack.pop();
    ans += i - j - 1;
  }
  stack.push([inputs[i], i]);
}

while (stack.length != 0) {
  const [_, j] = stack.pop();
  ans += N - j;
}

console.log(ans);
